### PR TITLE
Add Feature Model Cloning

### PR DESCRIPTION
**Pull request by s9latimm**
- Adds deep cloning for `FeatureModel` and related data structures (`Constraint`, `FeatureSourceRange`). 
- Reworks `getParentFeature` replacing `getParent` with a template in `FeatureTreeNode`.

### CODE REVIEW

File: include/vara/Feature/Constraint.h [hunk 1]
```diff
@@ -113,6 +115,12 @@ class OrConstraint : public BinaryConstraint, public BooleanConstraint {
       : BinaryConstraint(ConstraintKind::CK_OR, std::move(LeftOperand),
                          std::move(RightOperand)) {}
 
+  std::unique_ptr<Constraint> clone() override {
+    return std::make_unique<OrConstraint>(
+        std::move(this->getLeftOperand()->clone()),
+        std::move(this->getRightOperand()->clone()));
```
Comment (vulder) [inline include/vara/Feature/Constraint.h]:
```suggestion
        this->getLeftOperand()->clone(),
        this->getRightOperand()->clone());
```
The returned value is already an rvalue

File: include/vara/Feature/Constraint.h [hunk 2]
```diff
@@ -130,6 +138,12 @@ class XorConstraint : public BinaryConstraint, public BooleanConstraint {
       : BinaryConstraint(ConstraintKind::CK_XOR, std::move(LeftOperand),
                          std::move(RightOperand)) {}
 
+  std::unique_ptr<Constraint> clone() override {
+    return std::make_unique<XorConstraint>(
+        std::move(this->getLeftOperand()->clone()),
+        std::move(this->getRightOperand()->clone()));
```
Comment (vulder) [inline include/vara/Feature/Constraint.h]:
```suggestion
        this->getLeftOperand()->clone(),
        this->getRightOperand()->clone());
```

File: include/vara/Feature/Constraint.h [hunk 3]
```diff
@@ -147,6 +161,12 @@ class AndConstraint : public BinaryConstraint, public BooleanConstraint {
       : BinaryConstraint(ConstraintKind::CK_AND, std::move(LeftOperand),
                          std::move(RightOperand)) {}
 
+  std::unique_ptr<Constraint> clone() override {
+    return std::make_unique<AndConstraint>(
```
Comment (vulder) [inline include/vara/Feature/Constraint.h]: etc...

File: include/vara/Feature/Constraint.h [hunk 4]
```diff
@@ -481,10 +587,14 @@ class Feature;
 class PrimaryFeatureConstraint : public PrimaryConstraint {
 public:
   PrimaryFeatureConstraint(std::variant<Feature *, std::unique_ptr<Feature>> FV)
-      : PrimaryConstraint(ConstraintKind::CK_FEATURE), FV(std::move(FV))  {}
+      : PrimaryConstraint(ConstraintKind::CK_FEATURE), FV(std::move(FV)) {}
 
   [[nodiscard]] Feature *getFeature() const;
 
+  std::unique_ptr<Constraint> clone() override {
+    return std::make_unique<PrimaryFeatureConstraint>(this->getFeature());
```
Comment (vulder) [inline include/vara/Feature/Constraint.h]: Is this feature ptr later replaced?

File: include/vara/Feature/Feature.h [hunk 5]
```diff
@@ -50,18 +50,6 @@ class Feature : public FeatureTreeNode {
 
   [[nodiscard]] bool isOptional() const { return Opt; }
 
-  /// Search parent feature in tree structure -- this may not exist or nullptr
-  /// if node is already root.
-  [[nodiscard]] Feature *getParentFeature() const {
```
Comment (vulder) [inline include/vara/Feature/Feature.h]: Why did you move this to the `TreeNodeClass`? As an abstraction this makes more sense on a Feature, with the move the TreeNode is now hard Tied to the Feature.

File: include/vara/Feature/Feature.h [hunk 6]
```diff
@@ -225,12 +213,13 @@ class BinaryFeature : public Feature {
   BinaryFeature(
       string Name, bool Opt = false,
       std::vector<FeatureSourceRange> Loc = std::vector<FeatureSourceRange>(),
-      Feature *Parent = nullptr, const NodeSetType &Children = {})
+      FeatureTreeNode *Parent = nullptr, const NodeSetType &Children = {})
```
Comment (vulder) [inline include/vara/Feature/Feature.h]: A gain, more abstraction breaks. I don't think we should rework this.

**Review by vulder — CHANGES_REQUESTED** [review]
Hmm, I think we should not break the Feature/FeatureTreeNode abstraction. Why did you mix these two to implement clone?

File: include/vara/Feature/Feature.h (hunk 5 above)
Comment (s9latimm) [reply include/vara/Feature/Feature.h]: `Relationship` now needs a look up to a parent feature too. I reworked this as a template in `FeatureTreeNode`.

File: include/vara/Feature/Constraint.h (hunk 4 above)
Comment (s9latimm) [reply include/vara/Feature/Constraint.h]: I reworked this part, so that this copy only holds a new shallow feature (name only), which later needs replacement (the `FeatureModelBuilder` used cloning a `FeatureModel` already does exactly that).

Also, added comment should explain the intended behavior.

File: include/vara/Feature/Feature.h (hunk 5 above)
Comment (vulder) [reply include/vara/Feature/Feature.h]: I fixed the Relationship with local cast handling, we don't really need the template.

File: unittests/Feature/FeatureModel.cpp [hunk 7]
```diff
@@ -13,6 +13,79 @@ TEST(FeatureModel, build) {
...
+  auto FM = B.buildFeatureModel();
+  assert(FM);
+
+  auto Clone = FM->clone();
+  assert(Clone);
+  FM.reset();
+
+  // NOLINTNEXTLINE
+  ASSERT_EXIT(EXPECT_TRUE(FM->getFeature("a")->isRoot()),
+              testing::KilledBySignal(SIGSEGV), ".*");
```
Comment (vulder) [inline unittests/Feature/FeatureModel.cpp:44]: Why do you need the EXPECT_TRUE inside the EXIT? Should the segfault not also occur without it?

**Review by vulder — APPROVED** [review]
